reactive/etcd.py

```diff
@@ -175,58 +145,190 @@ def install_etcd():
-        cluster_data['token'] = etcd_helper.cluster_token()
-        cluster_data['cluster_state'] = 'existing'
-        cluster_data['cluster'] = etcd_helper.cluster_string()
-        cluster_data['leader_address'] = unit_get('private-address')
-        # Actually broadcast that gnarly dictionary
-        leader_set(cluster_data)
-        # leader assumes new? seems to work.
-        cluster_data['cluster_state'] = 'new'
-    else:
-        status_set('maintenance', 'registering unit with etcd-leader')
-        cluster_data['token'] = leader_get('token')
-        cluster_data['cluster_state'] = leader_get('cluster_state')
-        cluster_data['cluster'] = etcd_helper.cluster_string()
-        cluster_data['leader_address'] = leader_get('leader_address')
-        # self registration provided via the helper class
-        etcd_helper.register(cluster_data)
+@when('etcd.ssl.placed')
+@when_not('leadership.is_leader')
+@when_not('etcd.registered')
+def register_node_with_leader():
+    '''
+    Control flow mechanism to perform self registration with the leader.
+
+    Before executing self registration, we must adhere to the nature of offline
+    static turnup rules. If we find a GUID in the member list without peering
+    information the unit will enter a race condition and must wait for a clean
+    status output before we can progress to self registration.
+    '''
+    # We're going to communicate with the leader, and we need our bootstrap
+    # startup string once.. TBD after that.
+    etcdctl = EtcdCtl()
+    bag = EtcdDatabag()
+    # Assume a hiccup during registration and attempt a retry
+    if bag.cluster_unit_id:
+        bag.cluster = bag.registration_peer_string
+        render('defaults', '/etc/default/etcd', bag.__dict__)
+        host.service_restart('etcd')
+        time.sleep(2)
+
+    peers = etcdctl.member_list(leader_get('leader_address'))
+    for unit in peers:
+        if 'client_urls' not in peers[unit].keys():
+            # we cannot register. State not attainable.
+            msg = 'Waiting for unit to complete registration'
+            status_set('waiting', msg)
+            return
+
+    if not bag.cluster_unit_id:
+        bag.leader_address = leader_get('leader_address')
+        resp = etcdctl.register(bag.__dict__)
+        if resp and 'cluster_unit_id' in resp.keys() and 'cluster' in resp.keys():
+            bag.cache_registration_detail('cluster_unit_id',
+                                          resp['cluster_unit_id'])
+            bag.cache_registration_detail('registration_peer_string',
+                                          resp['cluster'])
+
+            bag.cluster_unit_id = resp['cluster_unit_id']
+            bag.cluster = resp['cluster']
+
+    render('defaults', '/etc/default/etcd', bag.__dict__)
+    host.service_restart('etcd')
+    time.sleep(2)
+
+    # Check health status before we say we are good
+    etcdctl = EtcdCtl()
```

Comment:
Why is it necessary to re-create an EtcdCtl object here? A comment to why a new object is necessary is more helpful than the general comment (which I am totally guilty of myself).